packages/server/lib/browsers/webkit.ts

```diff
@@ -0,0 +1,82 @@
+import Debug from 'debug'
+import { EventEmitter } from 'events'
+import type playwright from 'playwright-webkit'
+import type { Browser, BrowserInstance } from './types'
+import type { Automation } from '../automation'
+import { WebkitAutomation } from './webkit-automation'
+
+const debug = Debug('cypress:server:browsers:webkit')
+
+let wkAutomation: WebkitAutomation | undefined
+
+export async function connectToNewSpec (browser: Browser, options, automation: Automation) {
+  if (!wkAutomation) throw new Error('connectToNewSpec called without wkAutomation')
+
+  automation.use(wkAutomation)
+  await options.onInitializeNewBrowserTab()
+  await wkAutomation.reset(options.url)
+}
+
+export async function open (browser: Browser, url, options: any = {}, automation: Automation): Promise<BrowserInstance> {
+  // resolve pw from user's project path
+  const pwModulePath = require.resolve('playwright-webkit', { paths: [process.cwd()] })
+  const pw = require(pwModulePath) as typeof playwright
+
+  const pwBrowser = await pw.webkit.launch({
+    proxy: {
+      server: options.proxyServer,
+    },
+    downloadsPath: options.downloadsFolder,
+    headless: browser.isHeadless,
+  })
+
+  let pwPage: playwright.Page
+
+  async function resetPage (_url) {
+    // new context comes with new cache + storage
+    const newContext = await pwBrowser.newContext({
+      ignoreHTTPSErrors: true,
+    })
+    const oldPwPage = pwPage
+
+    pwPage = await newContext.newPage()
+
+    let promises: Promise<any>[] = []
+
+    if (oldPwPage) promises.push(oldPwPage.context().close())
+
+    if (_url) promises.push(pwPage.goto(_url))
+
+    if (promises.length) await Promise.all(promises)
+
+    return pwPage
+  }
+
+  pwPage = await resetPage(url)
+
+  wkAutomation = new WebkitAutomation(resetPage, pwPage)
```

Comment:
I notice we pass in a `resetPage` function to the constructor then use it within a variable we assign in the constructor. Since this function is already webkit specific, is there a compelling reason to pass all this logic as a function?

Eg - should we just move the entire body of `resetPage` ([this part](https://github.com/cypress-io/cypress/pull/15533/files#diff-d79130e57e24b51b5f19ffab413c34c1370b0524a971192d366af18838477ad7R35-R53)) to be declared inside of the `WebkitAutomation` class, and just pass it what's necessary (in this case, that would be the `url` and `pwBrowser`, I guess?)

Just an idea - this logic feels like it belongs to `WebkitAutomation` more than it does in a `browsers/webkit` module.

Reply:
Agreed, this became more obvious as I added more per-context setup stuff. I'm refactoring this in https://github.com/cypress-io/cypress/pull/23232